### ISSUE TITLE
SafeMath is not very safe when you overflow inputs (obviously).

### ISSUE THREAD

**Issue by destenson** [OpenZeppelin/openzeppelin-solidity#1505]
My understanding of SafeMath is that it's supposed to prevent invalid calculations due to negative inputs.

**💻 Environment**

Any smart contract that depends on SafeMath for valid calculations.

**📝 Details**

For each of the 4 math functions, there are invalid inputs that give invalid outputs without throwing.

**🔢 Code to reproduce bug**

EDIT: I'm wrong about these examples. In my excitement to share what I observed when the inputs had been overflowed before giving them to the library, I chose the simplest examples I could find, without realizing that they're obviously not correct.

I would think these should all throw and don't: 
```
add(uint(-1),1);
div(uint(-1),1);
mul(uint(-1),1);
sub(uint(-1),1);
```

I didn't test exhaustively, or work out the math, but my intuition says each of them can be exploited to output almost any desired value.

I think any contracts that use the vulnerable code should be deprecated immediately.

**Comment by destenson** [comment]
I would be remiss if I didn't post a fix to these. These are correct and probably use significantly less gas than the faulty versions.

```
    function add(uint a, uint b) internal pure returns(uint y) {
        y = a + b;
        require(int(y) >= 0);
    }
    
    function div(uint a, uint b) internal pure returns (uint y) {
        y = a / b;
        require(int(y) >= 0);
    }
    
    function mul(uint a, uint b) public pure returns(uint y) {
        y = a * b;
        require(int(y) >= 0);
    }
    
    function sub(uint a, uint b) public pure returns (uint y) {
        y = a - b;
        require(int(y) >= 0);
    }
```

A word to the wise: please don't write code to do what the compiler is supposed to do for you. If it seems like you're reinventing the wheel, you are.

**Comment by redsquirrel** [comment]
Do you understand that `uint(-1)` returns 115792089237316195423570985008687907853269984665640564039457584007913129639935? 
And that by definition, casting an `uint` to an `int` will *always* result in an int that is `>= 0`.

**Comment by destenson** [comment]
Yes, I do understand that. What is the point of this library if not to protect from that?

**Comment by destenson** [comment]
The fault is not with the library, _per se_, though it can be improved & I'm doing that today. The real issue is using types that can overflow & assuming they won't.

**Comment by destenson** [comment]
To follow up, this issue was caused by inadvertent overflow on my part before passing to the functions without realizing it. Then in my excitement to share what I found, I chose the simplest test cases I could find without realizing that they were not even behaving incorrectly.

Ultimately, my issue is with Solidity. EVM operations overflow & that's fine. But as a high-level language, Solidity should only allow variables to overflow when they are intended to and generate code that reverts upon overflow in the default case. I suspect that would involve the addition of new overflowable types that work the same as existing ones, and replacing the existing ones with types that generate code that reverts upon overflow.

Thus, this issue is closed, and I apologize for bringing it up so prematurely.
